flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1358,4 +1358,19 @@ public void actionsUnavailable() throws InterruptedException {
 			mailbox.putMail(actionUnavailableLetter);
 		}
 	}
+
+	private class TimerInvocationContext implements SystemProcessingTimeService.ScheduledCallbackExecutionContext {
+		@Override
+		public void invoke(ProcessingTimeCallback callback, long timestamp) throws InterruptedException {
+			mailbox.putMail(() -> {
+				synchronized (getCheckpointLock()) {
```

Comment:
Does here still have to require the checkpointLock? I think it has synchronized by mailbox.

Reply:
Hi @Aitozi, thank you for the review!

Personally, I'm not sure if this lock can be dropped at current stage (I'm yet trying to understand the runtime and which things happen in different tasks in parallel).

I've preserved it to keep old semantics when the callback was invoked in the timer's service thread pool (just to be "safe"). My reasoning was that some implementations of `onProcessingTime` may be missing proper synchronization lock (hence, the original reason why the timer service had it in the first place) **and** at the current stage, there are still some additional threads that may access tasks' state and make some operations (in particular methods related to checkpoints handling).

Although `SourceStreamTask` already has such lock around invocation in it's own mailbox loop, but the base `StreamTask` doesn't have it (the base class make invoke some letters before reaching the default action).

Stefan also thinks that this lock is not needed anymore. I can try to read the code again more thoroughly, to check if `onProcessingTime` is either
 * guaranteed to be executed under `SourceStreamTask.defaultAction`;
 * use locking internally, as implementation detail;
 * don't need locking at all.

Regarding timer cancellation, my thought is that it isn't guaranteed anyway: the code that use those timers, may cancel the timer, but it didn't have any guarantees if the timer has already fired and the callback was already invoked.
So, to support cancellation of `ProcessingTimeCallback`s that are already in the mailbox, would be a slight optimization (by stretching cancellation time window untill the callback is in the mailbox). But not having it, should be not worse situation as before, imo.

Reply:
1. ~~Regarding this checkpointLock, I think after this PR all timer's `onProcessingTime` will be executed in the main thread without multi-thread execution.~~ After double thought, I think you are right now.., because the checkpoint path has not been refactor to use mailbox to synchronize. The lock may be need to be removed by an individual PR later , i think.
2. Regarding timer cancellation, i want to hear some suggestion from Stefan.

Reply:
@1u0 I think there might be a missunderstanding. In our discussion I said that we might need to take care of `notifyCheckpointComplete` first before we can drop the lock.

Reply:
It is possible that there is some misunderstanding.
IRC, the discussion started in context of this PR only (not involving any future changes in other tickets). In particular, if the lock here is still needed (taking into account that there is a new lock in `SourceStreamTask` added in #8442).

Reply:
For this PR, I think it is still needed. After all checkpoint related messages go through the mailbox as well, I feel pretty sure that it is no longer needed.